RFC: propose helper script interface for writing rules

Ella-0 opened the pull request:
Proposes the use of a helper script to allow setting udev properties as a replacement for udev rules.  For example setting `ACP_IGNORE=1` on the visense pcm in apple silicon macs.
```sh
#!/bin/sh -e

if printf '%s\n' "$SYSNAME" | grep '^pcmC.*D2c$'
then
	printf 'ACP_IGNORE=1\n'
fi
```

Ella-0 committed:
Allow invoking helper script as a replacement for udev rules

illiliti commented on udev_device.c:
```diff
@@ -540,6 +545,49 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    }
+    snprintf(ptr, sizeof(helper_path), "%s", helper_path);
+
+    FILE *file = popen(cmd, "r");
```
`popen` is rather unsafe due to shell invocation. `posix_spawn` is better alternative, albeit it requires more code.

illiliti commented on udev_device.c:
```diff
@@ -540,6 +545,49 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        size += strlen(name) + 2 + strlen(value) + 2;
+    }
+
+    char helper_path[] = "/sbin/libudev-zero-rules-helper";
```
Put this string as a macro at top of the file.

illiliti reviewed (commented):
Thanks for working on this! I like the idea of passing control of properties to a helper script. It's much better option than adding special handling for properties.

xplshn commented:
@illiliti Will this not be merged into master/main?

illiliti commented:
> @illiliti Will this not be merged into master/main?

I like the idea, so it is highly likely I'll merge this. Just need to sort out issues with code.

Ella-0 commented:
Addressed all comments and added an example rules script to contrib with the `SOUND_INITIALIZED` and `ACP_IGNORE` stuff. Note: this on its own will not fix pipewire since pipewire still depends on the enumeration of parent devices

illiliti commented:
> Note: this on its own will not fix pipewire since pipewire still depends on the enumeration of parent devices

One could access parent device via `$SYSPATH/../` or `$SYSPATH/device/`, so unless I'm missing something it is indeed possible to fix pipewire.

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+
+    char *argv[] = { RULES_HELPER_NAME, NULL };
+
+    if (posix_spawnp(&pid, argv[0], &actions, NULL, argv, envp)) {
```
Use `posix_spawn`(no p). We don't need `PATH` traversal.

Ella-0 replied:
Where's the most sensible path to put the script?

illiliti replied:
Depends on a distro. Some prefer /usr/libexec, some /usr/lib, some /sbin and so on. Perhaps the most sensible path is no path? Let distro maintainers decide for themselves?

Ella-0 replied:
guess that means adding a way to configure it in the Makefile

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+
+    pid_t pid;
+
+    char *argv[] = { RULES_HELPER_NAME, NULL };
```
const

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+
+    int out_pipe[2];
+    posix_spawn_file_actions_t actions;
+    pipe(out_pipe);
```
Check error

Ella-0 commented:
> > Note: this on its own will not fix pipewire since pipewire still depends on the enumeration of parent devices
> 
> One could access parent device via `$SYSPATH/../` or `$SYSPATH/device/`, so unless I'm missing something it is indeed possible to fix pipewire.

The problem with pipewire is it enumerates the following devices
```
/sys/devices/platform/sound/sound/card0/pcmC0D0p
/sys/devices/platform/sound/sound/card0/pcmC0D1p
/sys/devices/virtual/sound/timer
/sys/devices/platform/sound/sound/card0/pcmC0D0c
/sys/devices/platform/sound/sound/card0/pcmC0D2c
/sys/devices/platform/sound/sound/card0/controlC0
```

however the following is intended. card0 is missing. this is due to scan_devices working on `/sys/dev/{char,block}` but `card?` does not have an associated character or block device so isn't enumerated. 

```
/sys/devices/platform/sound/sound/card0
/sys/devices/platform/sound/sound/card0/pcmC0D0p
/sys/devices/platform/sound/sound/card0/pcmC0D1p
/sys/devices/virtual/sound/timer
/sys/devices/platform/sound/sound/card0/pcmC0D0c
/sys/devices/platform/sound/sound/card0/pcmC0D2c
/sys/devices/platform/sound/sound/card0/controlC0
```

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        count += 1;
+    }
+
+    char **envp = malloc((count + 1) * sizeof(char *));
```
malloc is unnecessary because the number of properties is [hardcoded to 64](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/include/linux/kobject.h#n32)

Ella-0 replied:
Are there any existing macros for these two that I should be using?

illiliti replied:
None I'm aware of.

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    }
+
+    char **envp = malloc((count + 1) * sizeof(char *));
+    char *env = malloc(size);
```
Same as above. [buffer size is hardcoded](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/include/linux/kobject.h#n33)

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    char *argv[] = { RULES_HELPER_NAME, NULL };
+
+    if (posix_spawnp(&pid, argv[0], &actions, NULL, argv, envp)) {
+        perror("Could not launch properties helper");
```
We cannot log errors like that because we don't control stderr/stdout(we are library after all)

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+
+    int status;
+    waitpid(pid, &status, 0);
+    close(out_pipe[0]);
```
fclose

illiliti commented:
Ah! Right. Forgot about this one. Well, we could switch our enumeration approach to appease pipewire. Let's keep it for later.

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+         entry = udev_list_entry_get_next(entry)) {
+
+        const char *name = entry->name;
+        const char *value = entry->value?entry->value:"";
```
Add spaces between operators

illiliti commented on udev_device.c:
```diff
@@ -540,6 +551,76 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    char **envp = malloc((count + 1) * sizeof(char *));
+    char *env = malloc(size);
+    char *ptr = env;
+    unsigned i = 0;
```
size_t

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+
+    int out_pipe[2];
+
+    if (pipe(out_pipe)) {
```
Switch to pipe2 and set O_CLOEXEC

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    if (posix_spawn(&pid, argv0, &actions, NULL, argv, envp)) {
+        return -1;
+    }
+
```
Put posix_spawn_file_actions_destroy somewhere here

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        const char *value = entry->value ? entry->value : "";
+
+        // '{name}={value}\0'
+        size_t max_size = strlen(name) + 1 + strlen(value) + 1;
```
Add bounds checking here. Simple ```assert()``` would be enough

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        size_t max_size = strlen(name) + 1 + strlen(value) + 1;
+        snprintf(ptr, max_size, "%s=%s", name, value);
+
+        envp[i] = ptr;
```
bounds checking

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        i++;
+    }
+
+    envp[i] = NULL;
```
bounds checking

illiliti commented on udev_device.c:
```diff
@@ -540,6 +559,70 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    posix_spawn_file_actions_init(&actions);
+    posix_spawn_file_actions_addclose(&actions, out_pipe[0]);
+    posix_spawn_file_actions_adddup2(&actions, out_pipe[1], 1);
+    posix_spawn_file_actions_addclose(&actions, out_pipe[1]);
```
I wonder if we should no-op(or close?) stdin/stderr using /dev/null to prevent child process from inheriting parent's stdin/stderr

illiliti commented on Makefile:
```diff
@@ -3,11 +3,13 @@
 		  -Wall -Wextra -Wpedantic -Wmissing-prototypes -Wstrict-prototypes \
-		  -Wno-unused-parameter
+		  -Wno-unused-parameter \
+		  -DRULES_HELPER_PATH=\"${LIBEXECDIR}/libudev-zero-rules-helper\"
```
It'd be better if RULES_HELPER_PATH had its own make macro, so user could set it on the command line when building just like we set PREFIX and similar...

illiliti commented on Makefile:
```diff
@@ -3,11 +3,14 @@
+LIBEXECDIR = ${PREFIX}/libexec
 INCLUDEDIR = ${PREFIX}/include
 PKGCONFIGDIR = ${LIBDIR}/pkgconfig
+RULES_HELPER_PATH=${LIBEXECDIR}/libudev-zero-rules-helper
```
space

illiliti commented on udev_device.c:
```diff
@@ -540,6 +561,85 @@ static void set_properties_from_props(struct udev_device *udev_device)
+        return -1;
+    }
+
+    int null_fd = open("/dev/null", O_CLOEXEC | O_RDWR);
```
this could be simplified using `posix_spawn_file_actions_addopen`, e.g:
```c
posix_spawn_file_actions_addopen(&actions, STDIN_FILENO, "/dev/null", O_RDONLY, 0);
posix_spawn_file_actions_addopen(&actions, STDERR_FILENO, "/dev/null", O_WRONLY, 0);
```

illiliti commented on udev_device.c:
```diff
@@ -540,6 +561,85 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    char *argv[] = { argv0, NULL };
+
+    if (posix_spawn(&pid, argv0, &actions, NULL, argv, envp)) {
+        return -1;
```
close out_pipe and destroy actions

illiliti replied:
```c
int ret = posix_spawn(...);
posix_spawn_file_actions_destroy(&actions);
close(out_pipe[1]);
if (ret != 0) {
    close(out_pipe[0]);
    return -1;
}
```

illiliti replied:
You are still leaking `out_pipe` and `actions` here. Consider adopting the above code to fix the leak.

illiliti commented on udev_device.c:
```diff
@@ -540,6 +561,85 @@ static void set_properties_from_props(struct udev_device *udev_device)
+    FILE *file = fdopen(out_pipe[0], "r");
+
+    if (!file) {
+        return -1;
```
close out_pipe and waitpid(how to prevent zombie otherwise? posix_spawn does not have POSIX_SPAWN_NOZOMBIE flag like [qnx has](http://www.qnx.com/developers/docs/7.0.0/index.html#com.qnx.doc.neutrino.lib_ref/topic/p/posix_spawn.html))

Ella-0 replied:
I'm not 100% on my solution. I've never had to think about this properly before. Can you confirm its correctness?

illiliti replied:
Due to an early return there is a leak of out_pipe and zombie process. Make sure all resources are free'd before returning, e.g:

```c
FILE *file = fdopen(out_pipe[0], "r");
if (!file) {
    // fix fd leak
    close(out_pipe[0]);
    // fix zombie process leak
    waitpid(pid, NULL, 0);
    return -1;
}

set_properties_from_file(udev_device, file);
fclose(file);
waitpid(pid, NULL, 0);
return 0;
```

Ella-0 committed:
contrib: add example rules script

Ella-0 commented on udev_device.c:
```diff
@@ -540,6 +561,85 @@
+
+    int out_pipe[2];
+
+    if (pipe2(out_pipe, O_CLOEXEC)) {
```
uhhh this is in unistd.h according to docs I think? do correct me if I'm wrong @illiliti

illiliti replied:
Set `GNU_SOURCE` in the makefile instead of `XOPEN_SOURCE` in order for `pipe2` to be avaiable.

xplshn commented:
@illiliti any plans to merge this or work on it?

illiliti commented:
Hopefully yes

WhyNotHugo commented:
Do I understand correctly that every script would run for every event received by every process subscribed to events?

Ella-0 commented:
Hi sorry about the delay on this. I am very very busy at the moment.
